src/IO/S3/copyS3File.h

```diff
@@ -21,7 +21,12 @@ using CreateReadBuffer = std::function<std::unique_ptr<SeekableReadBuffer>()>;
 /// The same functionality can be done by using the function copyData() and the classes ReadBufferFromS3 and WriteBufferFromS3
 /// however copyS3File() is faster and spends less network traffic and memory.
 /// The parameters `src_offset` and `src_size` specify a part in the source to copy.
+///
+/// Note, that it tries to copy file using native copy (CopyObject), but if it
+/// has been disabled (with settings.allow_native_copy) it is fallbacks to
+/// read-write copy (copyDataToS3File()).
 void copyS3File(
+    const CreateReadBuffer & create_read_buffer,
```

Comment:
How? It does not have a disk to read it that can be encrypted or not (for BACKUPs) and also it may have different credentials then the destination disk.

Reply:
How - again, see `CopyFileHelper::processCopyRequest()`.

No, it doesn't need a disk and also it doesn't have to check if the disk is encrypted or not, because it's already checked (https://github.com/ClickHouse/ClickHouse/blob/c7ae150ffb8c9e1c28262e8512074ffa378bd71f/src/Backups/BackupIO_S3.cpp#L197)

Reply:
But still, in this case copy can fail in source and destination bucket requires different credentials, right?
And `getSourceObjectReadBuffer` suffers from the same problem

Reply:
If the endpoints differ are not the same then it won't come to this point anyway (see [check](https://github.com/ClickHouse/ClickHouse/blob/c7ae150ffb8c9e1c28262e8512074ffa378bd71f/src/Backups/BackupIO_S3.cpp#L197)).

Is it possible that different credentials should be used for the same endpoint for reading from a disk and writing a backup?

Reply:
>If the endpoints are not the same then it won't come to this point anyway (see [check](https://github.com/ClickHouse/ClickHouse/blob/c7ae150ffb8c9e1c28262e8512074ffa378bd71f/src/Backups/BackupIO_S3.cpp#L197)).

Yep. you are right looks safe then.

> Is it possible that different credentials should be used for the same endpoint for reading from a disk and writing a backup?

I guess it is, but very unlikely, let's address this separately anyway (to not pack to much changes)